### ISSUE TITLE
Unsupported destructering with dynamic keys in flow. Bug or just not possible?

### ISSUE THREAD

**Issue by meandmax** [facebook/flow#1711]
I found a lot of issues on destructering but not exactly this case and there I made this issue, anybody an idea?

Thanks! ;)

http://tryflow.org/#9236729c4f701bacd27213fb32ce8bce

```js
export const a = 'hast';
export const b = 'bier';


const hans = {
    hast: {
        one: 'two',
        three: 'four'
    },
    bier: {
        five: 'two',
        six: 'four'
    }
};

const {
    [a]: { // unsupported destructering
        one,
        three
    },
    [b]: { // unsupported destructering
        five,
        six
    }
} = hans;
```

**Comment by jeffmo** [comment]
This is indeed a bug in Flow! Sorry about that -- hopefully we'll have this fixed soon

**Comment by mroch** [comment]
for whoever implements this, note that `const {[a]} = hans` is invalid. the `":" Initializer` is required.

**Comment by meandmax** [comment]
@mroch could you explain this `":" Initializer` in one sentence or give us a link, because Im not sure what you mean?

Thanks

**Comment by mroch** [comment]
`const {[a]: foo} = bar;` is valid, because it creates a `foo` binding as if you had written `const foo = bar[a]`. `const {[a]} = bar;` is invalid, because it would be like writing `const <value of a> = bar[a]`, and you can't have variable variable names.

in the ES6 grammar, it's a literal colon `":"` followed by an `Initializer`.

**Comment by mroch** [comment]
specifically, http://www.ecma-international.org/ecma-262/6.0/#sec-destructuring-binding-patterns

_BindingProperty_ is a _SingleNameBinding_ or _PropertyName_ `:` _BindingElement_.
_PropertyName_ is either a literal property name or a computed property name, but in either case requires a `:` plus _BindingElement_, and _BindingElement_ is a _BindingIdentifier_.

(realized I mispoke above, I meant `Identifier`, not `Initializer`)

**Comment by kevinSuttle** [comment]
A more simple example:
```js
 const Heading = ({ type, children }) => {
                      ^^^^^^^^^^^^^^^^^^ destructuring. Missing annotation
```

**Comment by samwgoldman** [comment]
The original issue is resolved by a70d613f21b66d9e56a15fca9c7a5bdb8dd54bce. @kevinSuttle, your issue is different. The missing annotation error is correct, because all exported functions need to be annotated.
